subsys/usb/usb_c/usbc_tc_common.c

```diff
@@ -0,0 +1,264 @@
+ */
+void tc_subsys_init(const struct device *dev)
+{
+	struct usbc_port_data *data = dev->data;
+	struct tc_sm_t *tc = data->tc;
+
+	/* Save the port device object so states can access it */
+	tc->dev = dev;
+
+	/* Initialize the state machine */
+	smf_set_initial(SMF_CTX(tc), &tc_states[TC_DISABLED_STATE]);
+}
+
+/**
+ * @brief Runs the Type-C layer
+ */
+void tc_run(const struct device *dev, const int32_t dpm_request)
+{
+	struct usbc_port_data *data = dev->data;
+	const struct device *tcpc = data->tcpc;
+	struct tc_sm_t *tc = data->tc;
+
+	/* These requests are implicitly set by the Device Policy Manager */
+	if (dpm_request == PRIV_PORT_REQUEST_START) {
+		data->tc_enabled = true;
+	} else if (dpm_request == PRIV_PORT_REQUEST_SUSPEND) {
+		data->tc_enabled = false;
+		tc_set_state(dev, TC_DISABLED_STATE);
+	}
+
+	switch (data->tc_sm_state) {
+	case SM_PAUSED:
+		if (data->tc_enabled == false) {
+			break;
+		}
+	/* fall through */
+	case SM_INIT:
+		/* Initialize the Type-C layer */
+		tc_init(dev);
+		data->tc_sm_state = SM_RUN;
+	/* fall through */
+	case SM_RUN:
+		if (data->tc_enabled == false) {
+			pd_enable(dev, false);
+			data->tc_sm_state = SM_PAUSED;
+			break;
+		}
+
+		/* Sample CC lines */
+		tcpc_get_cc(tcpc, &tc->cc1, &tc->cc2);
+
+		/* Detect polarity */
+		tc->cc_polarity = (tc->cc1 > tc->cc2) ?
+			TC_POLARITY_CC1 : TC_POLARITY_CC2;
+
+		/* Execute any asyncronous Device Policy Manager Requests */
+		if (dpm_request == REQUEST_TC_ERROR_RECOVERY) {
+			/* Transition to Error Recovery State */
+			tc_set_state(dev, TC_ERROR_RECOVERY_STATE);
+		} else if (dpm_request == REQUEST_TC_DISABLED) {
+			/* Transition to Disabled State */
+			tc_set_state(dev, TC_DISABLED_STATE);
+		}
+
+		/* Run state machine */
+		smf_run_state(SMF_CTX(tc));
+	}
+}
+
+/**
+ * @brief Checks if the TC Layer is in an Attached state
+ */
+bool tc_is_in_attached_state(const struct device *dev)
+{
+	return (tc_get_state(dev) == TC_ATTACHED_SNK_STATE);
+}
+
+/**
+ * @brief Initializes the Type-C layer
+ */
+static void tc_init(const struct device *dev)
+{
+	struct usbc_port_data *data = dev->data;
+	struct tc_sm_t *tc = data->tc;
+
+	/* Initialize the timers */
+	usbc_timer_init(&tc->tc_t_error_recovery, TC_T_ERROR_RECOVERY_SOURCE_MIN_MS);
+	usbc_timer_init(&tc->tc_t_cc_debounce, TC_T_CC_DEBOUNCE_MAX_MS);
+	usbc_timer_init(&tc->tc_t_rp_value_change, TC_T_RP_VALUE_CHANGE_MAX_MS);
+
+	/* Clear the flags */
+	tc->flags = ATOMIC_INIT(0);
+
+	/* Initialize the TCPC */
+	tcpc_init(data->tcpc);
+
+	/* Initialize the state machine */
+	/*
+	 * Start out in error recovery state so the CC lines are opened for a
+	 * short while if this is a system reset.
+	 */
+	tc_set_state(dev, TC_ERROR_RECOVERY_STATE);
+}
+
+/**
+ * @brief Sets a Type-C state
+ */
+void tc_set_state(const struct device *dev, const enum tc_state_t state)
+{
+	struct usbc_port_data *data = dev->data;
+	struct tc_sm_t *tc = data->tc;
+
+	smf_set_state(SMF_CTX(tc), &tc_states[state]);
```

Comment:
Check that `state < ARRAY_SIZE(tc_states)`.  This could potentially be a build time check because all callers to `tc_set_state()` use a build time constant.

```
#define TC_SET_STATE(dev, state) \
    BUILD_ASSERT(state < ARRAY_SIZE(tc_states)); \
    smf_set_state(SMF_CTX(tc), &tc_states[state]);
```

Reply:
ARRAY_SIZE(tc_states) causes an error

Reply:
__ASSERT works but BUILD_ASSERT causes the error